libs/wingsdk/src/target-sim/resource.ts

```diff
@@ -1,5 +1,308 @@
+  INFO = "info",
+  /** Warning level */
+  WARN = "warn",
+  /** Error level */
+  ERROR = "error",
+}
+
+/**
+ * Global identifier for `sim.Resource`.
+ */
+export const SIM_RESOURCE_FQN = "@winglang/sdk.sim.Resource";
+
+/**
+ * A backend for a simulated resource.
+ * @inflight `@winglang/sdk.sim.IResourceClient`
+ */
+export class Resource
+  extends StdResource
+  implements
+    IStdResource,
+    ISimulatorResource,
+    IInflightHost,
+    ISimulatorInflightHost
+{
+  private readonly permissions: Array<[IStdResource, string]> = [];
+  private readonly _env: Record<string, string> = {};
+  private readonly factory: IResourceFactory;
+  private readonly entrypoint: string;
+
+  constructor(scope: Construct, id: string, factory: IResourceFactory) {
+    super(scope, id);
+
+    Node.of(this).title = "Service";
+    Node.of(this).description = "A cloud service";
+
+    const assetName = ResourceNames.generateName(this, {
+      disallowedRegex: /[><:"/\\|?*\s]/g, // avoid characters that may cause path issues
+      case: CaseConventions.LOWERCASE,
+      sep: "_",
+    });
+
+    const workdir = App.of(this).workdir;
+    mkdirSync(workdir, { recursive: true });
+    const entrypoint = join(workdir, `${assetName}.cjs`);
+    this.entrypoint = entrypoint;
+
+    if (process.env.WING_TARGET) {
+      this.addEnvironment("WING_TARGET", process.env.WING_TARGET);
+    }
+
+    this.factory = factory;
+  }
+
+  public addPermission(resource: IStdResource, op: string): void {
+    this.permissions.push([resource, op]);
+  }
+
+  /** @internal */
+  public get _liftMap(): LiftMap {
+    return {
+      [ResourceInflightMethods.CALL]: [],
+    };
+  }
+
+  /**
+   * Obtain a token that can be used to reference an attribute of this
+   * resource that is only resolved once the resource is started in the simulator.
+   *
+   * If the token is used in inflight code or in the configuration of a simulated
+   * resource (e.g. as an environment variable), the relevant resource will
+   * automatically take a dependency on the resource the attribute belongs to.
+   *
+   * @param name The name of the attribute.
+   * @returns An attribute token.
+   */
+  public attrToken(name: string): string {
+    return simulatorAttrToken(this, name);
+  }
+
+  /** @internal */
+  public _preSynthesize(): void {
+    super._preSynthesize();
+
+    const inflightClient = this.factory._toInflight();
+    const code = `\
+        "use strict";
+        let $klass;
+        exports.start = async function(statedir) {
+          if ($klass) {
+            throw Error('resource already started');
+          }
+          const attrs = {};
+          const ctx = {};
+          ctx.statedir = () => statedir;
+          ctx.resolveAttr = (name, value) => attrs[name] = value;
+          ctx.log = (message, level) => {
+            if (!level) level = 'info';
+            console.log(level + ': ' + message);
+          };
+          const client = ${inflightClient};
```

Comment:
Definitely seems like you can pass `ctx` to the factory function instead of `onStart`